OnlineLibrary.BLL/Services/BookService.cs

```diff
@@ -31,75 +34,55 @@ public List<Book> GetAllBooks()
-            if (
-                    (authorId == null)
-                    && String.IsNullOrEmpty(name)
-                    && (inReserve == null || inReserve < 0 || inReserve > 1)
-                    && (inArchieve == null || inArchieve < 0 || inArchieve > 1)
-                 )
-            {
-                List<Book> allBooks =  GetAllBooks();
-                ExceptionHelper.Check<OLNotFound>(allBooks == null || !allBooks.Any(), "Books don't exist");
-                return allBooks;
-            }
+            int count = unitOfWork.BookRepository.GetAllBooksCount();
+            if (count == 0) return new PageOfEntities<Book>();
+            int skip = (paginationOptions.PageNumber - 1) * paginationOptions.PageSize;
+            PageOfEntities<Book> page = new PageOfEntities<Book>();
+            page.EntitiesCount = count;
 
-            List<Book> books;
-            // Проверяем первый случай когда заполнены поля "имя" и "номер автора".
-            if ((authorId != null) && !String.IsNullOrEmpty(name))
-            {
-                books = unitOfWork.BookRepository.FilterBooks((int)authorId, name);
-            }
-            // Заполнен только автор книги.
-            else if (authorId != null)
-            {
-                books = unitOfWork.BookRepository.FilterBooks((int)authorId);
-            }
-            // Заполнено только название книги.
-            else if (!String.IsNullOrEmpty(name))
+            if (count - skip >= 1)
             {
-                books = unitOfWork.BookRepository.FilterBooks(name);
+                page.Entities = unitOfWork.BookRepository.GetAllBooks(skip, paginationOptions.PageSize);
+                page.Page = paginationOptions.PageNumber;
             }
-            // Поля фильтрации пустые, получаем весь список.
             else
             {
-                books = unitOfWork.BookRepository.GetAllBooks();
+                page.Page = (count / paginationOptions.PageSize) + 1;
+                skip = (page.Page - 1) * paginationOptions.PageSize;
+                page.Entities = unitOfWork.BookRepository.GetAllBooks(skip, paginationOptions.PageSize);
             }
+            page.ElementsOnPage = page.Entities.Count();
+            return page;
+        }
 
-            ExceptionHelper.Check<OLNotFound>(books == null || !books.Any(), "There are no books matching this search");
-
-            // Выбраны поля резервации и архивации.
-            if (inReserve >= 0 && inReserve <= 1 && inArchieve >= 0 && inArchieve <= 1)
+        public List<Book> FilterBooks(int? authorId, string name, int? inReserve, int? inArchieve)
+        {
+            Expression<Func<Book, bool>> expr = PredicateBuilder.New<Book>(true);
+            if (authorId != null && authorId > 0)
+            {
+                expr = expr.And(b=>b.Authors.Contains(unitOfWork.AuthorRepository.GetAuthorsByIdList(new List<int>(){ (int)authorId}).First())); // ????????????
```

Comment:
Are you sure that it correct expression with _unitOfWork.AuthorRepository.GetAuthorsByIdList(new List<int>(){ (int)authorId})_ inside
Strange comment in the end of the line